### PR TITLE
Panzer periodicbcbbox

### PR DESCRIPTION
**Pull request by reuterb**
@trilinos/panzer

## Motivation

For large periodic meshes, eliminating redundant degrees of freedom is a severe bottleneck at runtime due to global communication.
This change significantly reduces global communication by using some of the bounding box search functionality from STK.

## Testing

An acceptance test of the STK functionality is in panzer/adapters-stk/test/stk_interface_test/tCoarseSearch.cpp
Unit tests for the refactor are in panzer/adapters-stk/test/periodic_bcs/periodic_search.cpp

### CODE REVIEW

**Review by rppawlo — CHANGES_REQUESTED** [review]
@reuterb - looks good! Nice job on the unit tests! Very thorough! Just a few minor changes requested.

File: packages/panzer/adapters-stk/test/periodic_bcs/periodic_search.cpp [hunk 1]
```diff
@@ -0,0 +1,2304 @@
...
+    std::sort(L_unique_exact.begin(),L_unique_exact.end());
+    std::sort(uniqueLeftIds.begin(),uniqueLeftIds.end());
+
+    TEST_ASSERT(L_unique_exact.size()==uniqueLeftIds.size());
+    size_t errLUnique = 0;
+    for (size_t i=0; i<L_unique_exact.size(); ++i) {
+      if (L_unique_exact[i] != uniqueLeftIds[i]) 
+        ++errLUnique;
+    }
+    TEST_ASSERT(errLUnique==0);
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/test/periodic_bcs/periodic_search.cpp]: I tried to cover most, if not all, of these types of macro issues.

File: packages/panzer/adapters-stk/test/periodic_bcs/periodic_search.cpp [hunk 2]
```diff
@@ -0,0 +1,2304 @@
...
+        for (size_t j=0; j<nPoints; ++j){
+           // zy becomes xy
+           double exact[3] = {(double)i,(double)j,0.};
+           auto ptShifted = zy[n].first.center();
+           ++n;
+           if ( (std::abs(exact[0]-ptShifted[0]) > 1e-14) || (std::abs(exact[1]-ptShifted[1]) > 1e-14) ||
+                (std::abs(exact[2]-ptShifted[2]) > 1e-14) ) ++xzY_err;
+        }
+     }
+     TEUCHOS_ASSERT(xzY_err==0);
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/test/periodic_bcs/periodic_search.cpp]: Same here, tried to cover most of these.

File: packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_Interface.cpp [hunk 3]
```diff
@@ -1885,7 +1886,20 @@ STK_Interface::getPeriodicNodePairing() const
         type = 2;
       else
         TEUCHOS_ASSERT(false);
+#ifdef PANZER_HAVE_STKSEARCH
+      // warning if we have the new flag on but dont have STKSEARCH enable?
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_Interface.cpp]: Done.

File: packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_Interface.cpp [hunk 4]
```diff
@@ -1872,10 +1872,11 @@ STK_Interface::getPeriodicNodePairing() const
    Teuchos::RCP<std::vector<std::pair<std::size_t,std::size_t> > > vec;
    Teuchos::RCP<std::vector<unsigned int > > type_vec = rcp(new std::vector<unsigned int>);
    const std::vector<Teuchos::RCP<const PeriodicBC_MatcherBase> > & matchers = getPeriodicBCVector();
+   const bool & searchFlag = getPeriodicSearchFlag();
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_Interface.cpp]: Yes, this was changed and its knock on effects

File: packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_PeriodicBC_Parser.cpp [hunk 5]
```diff
@@ -51,6 +51,8 @@ namespace panzer_stk {
 PeriodicBC_Parser::PeriodicBC_Parser()
    : countStr_("Count")
    , condPrefix_("Periodic Condition ")
+   , searchStr_("Use BBox Search")
+   , newSearch_(false) // TODO BWR swap this to change default search (see also STK_interface.hpp)
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/src/stk_interface/Panzer_STK_PeriodicBC_Parser.cpp]: Resolved per our discussion.

File: packages/panzer/adapters-stk/test/periodic_bcs/periodic_32bit_int_limit.cpp [hunk 6]
```diff
@@ -59,8 +59,11 @@ TEUCHOS_UNIT_TEST(periodic_bcs, 32_bit_int_limit)
     if (true) {
       RCP<Teuchos::ParameterList> pbc = rcp(new Teuchos::ParameterList);
       pbc->set("Count",2);
-      pbc->set("Periodic Condition 1","xz-all 1e-12,3D: top;bottom");
-      pbc->set("Periodic Condition 2","yz-all 1e-12,3D: left;right");
+      pbc->set("Periodic Condition 1","xz-all 1e-12: top;bottom");
+      pbc->set("Periodic Condition 2","yz-all 1e-12: left;right");
```
[the remark this answers is not in the document]
Comment (reuterb) [reply packages/panzer/adapters-stk/test/periodic_bcs/periodic_32bit_int_limit.cpp]: There was a bug. I proposed a fix in my latest commit.

**Review by rppawlo — APPROVED** [review]
Looks good to me! Thanks @reuterb !

**Pull request merged**